## Numpy Financial

In [43]:
import pandas as pd
import numpy as np
import numpy_financial as npf

O **valor presente** nada mais é que o **valor de um montante do futuro** trazido para o **momento atual**. Então, ele serve basicamente para analisar o **valor** do dinheiro no tempo, considerando as variações que acontecem.

In [87]:
# Valor Presente 

vf = -1000000
juros = 0.05
pagamento = 0

valor_presente = npf.pv(juros, 10, pmt=0, fv=vf)

In [88]:
valor_presente

613913.2535407591

In [89]:
#Valor Futuro

valor_futuro = npf.fv(juros, 10, pmt=pagamento, pv=-300000)
valor_futuro

488668.3880332326

Também chamado de **Valor Líquido Atual**, o **Valor Presente Líquido** é um método que consiste em trazer para a data zero todos os fluxos de caixa de um projeto de investimento e somá-los ao valor do investimento inicial, usando como taxa de desconto a Taxa Mínima de Atratividade (TMA) da empresa ou projeto.

O **VPL** é um dos métodos mais conhecidos quando o assunto é análise da viabilidade de projetos de investimento.  Com o cálculo do Valor Presente Líquido é possível fazer os ajustes, descontando as taxas de juros para **obter a verdadeira noção do valor do dinheiro no futuro**.

Por esse motivo, o VPL não faz apenas uma comparação do investimento com o retorno que dele se espera. Indo além disso, como o Valor Presente Líquido leva em consideração a valorização do capital ao longo do tempo, com ele calcula-se o ganho real de investimento.

**Resumindo: O VPL traz ao valor presente um valor futuro. Ele representa a diferença entre os recebimentos e os pagamentos de um projeto de investimento em valores monetários atuais (de hoje).**

In [110]:
fc = np.array([-380, 30, 50, 70, 90, 110, 130, 130, 130, 130])
i = 0.12
vpl_teste = npf.npv(i, fc)
vpl_teste.round(2)

80.14

## Interpretando os resultados

Calculando a VPL e obtido o valor de 80.14, pela definição de VPL, significa que o projeto consegue recuperar o investimento inicial (380.00), remunera também aquilo que teria sido ganho se o capital para esse investimento ($380.00) tivesse sido aplicado na TMA (12.00 porcento a.a) e ainda sobram, em valores monetários de hoje 80.14 (excesso de caixa). 

**VPL > 0 indica que o Projeto merece continuar sendo analisado.**


Exemplo Acima do Livro  **Decisões Financeiras e Análise de Investimento**.

In [139]:
#VPL anualizado
n = np.count_nonzero(fc) - 1

eq1 = i*(1 + i)**n
eq2 = ((1 + i)**n) - 1
eq = eq1/eq2
vpl_teste

vpl_anualizado = vpl_teste * eq
vpl_anualizado.round(2)

15.04

Enquanto que no VPL todos os valores do fluxo de caixa são concentrados na data zero, no VPLa o fluxo de caixa representativo do projeto de investimento é transformado em uma série uniforme. 

**VPLa** ou VPL Anulizado **>0 indica que o Projeto merece continuar sendo analisado**. 

## IBC - Indice Beneficio/Custo

O IBC é uma medida de quanto se espera ganhar por unidade de capital investido. A hipótese implicita no cálculo do IBC é que os recursos liberados ao longo da vida util do projeto sejam reinvestidos à taxa minima atratividade (TMA). 

Genericamente, o IBC nada mais é do que uma razão entre o Fluxo Esperado de Beneficios de um projeto e o Fluxo Esperado de Investimentos necessários para realizá-lo. Assim, o IBC pode ser calculado pela fórmula: 


$ 
IBC = \frac{\text{Valor Presente do Fluxo de Beneficios}}{\text{Valor Presente do Fluxo de Investimentos}}$

In [145]:
IBC = (-fc[0]+vpl_teste)/-fc[0]
IBC

1.2108820350722522

O valor de IBC igual a 1,21 significa para que, cada 1 unidade monetária imobilizado no projeto, espera-se retirar, após o horizonte de planejamento do projeto(no caso do exemplo são nove anos), 1,21 após expurgado o ganho que se teria, caso esse 1 tivesse sido aplicado na TMA. Pode se raciocinar em termos de uma rentabilidade real esperada de 21,01% em nove anos. Note-se que essa taxa não permite comparação imediata com a TMA(12% a.a) porquanto a mesma se refere a um período de nove anos. Uma alternativa é encontrar a taxa equivalente para o mesmo período da TMA. Essa alternativa apresentará a rentabilidade esperada do projeto para o mesmo período da TMA e será denominada ROIA (Retorno Adicional sobre o Investimento). 

**IBC > 1 indica que o Projeto merece continuar sendo analisado**. 

## ROIA

O ROIA é a melhor estimativa de rentabilidade para um projeto de investimento. Representa, em termos percentuais, a riqueza gerada pelo projeto. Assim, o ROIA, é o analogo percentual do conceito de Valor Economico Agregado (EVA).

In [152]:
ROIA = npf.rate(n, 0, -1, IBC).round(4)
ROIA*100

2.15

O projeto em específico apresenta um ROI de 2,15% ao ano além da TMA(12%a.a). Essa informação é a melhor estimativa de rentabilidade do projeto de investimento. É importante considerar que o capital disponivel para o investimento já teria por denificação, uma aplicação de baixo risco com retorno de 12% ao ano. A decisão agora se resume em discutir se vale a pena investir no projeto (assumir o risco do investimento) para se ter um adicional de ganho da ordem de ganho de 2,15% ao ano. 

## TIR

In [171]:
tir =  npf.irr(fc)
tir

0.1618652950710049

Para o projeto em Análise, com TMA igual a 12% a.a e ROIA igual a 2,15% a.a, tem que se a seguinte relação:


In [191]:
#Relação entre ROIA e Tma
tma = (1+i)*(1+ROIA)-1
tma.round(5)

0.14408

In [192]:
if tma < tir:
    print("Indica que há mais ganhos investindo no Projeto do que na TMA")
else:
    print("Indica que NÃO há mais ganhos investindo no Projeto do que na TMA")

Indica que há mais ganhos investindo no Projeto do que na TMA


In [194]:
#fcr - Fluxo de Caixa Resultante dos Reinvestimentos
fcr = npf.fv(tir,nper=9,pmt=0, pv=fc)
fcr[0].round(2)

1466.17

In [193]:
#fcr - Com a TMA
fcr = npf.fv(tma,nper=9,pmt=0, pv=fc)
fcr[0].round(2)

1276.12

Por definição, a melhor alternativa de aplicação para os recursos liberados pelo projeto é a TMA (taca de aplicação com baixo grau de risco). Então a TIR somente poderá ser considerada como representativa da rentabilidade do projeto se houver uma coincidência de valores entre as taxas, isto é, TIR igual a TMA. 

Verifica-se que os recursos liberados pelo projeto forem aplicados à TMA, então o retorno sobre o capital investido será de 14.41% a.a. 

Resta agora, interpretar o significado desse valor. Defniniu-se, anteriormente o ROIA (2.15% a.a) como melhor estimativa de rentabilidade para o projeto em análise, após expurgado o efeito da TMA. Ora, sendo a TMA igual a 12% a.a e expurgando-se o efeito retorno 14,408% a.a chega-se a:

In [196]:
#ROIA
ROIA_1 = (1+tma) / (1 + i) - 1

ROIA_1

0.021500000000000075

Ou seja, o retorno de 14.408% a.a é o mesmo que o ROIA(2.15% a.a), porém com efeito cruzado da TMA. Dessa forma, a TIR, na melhor das hipóteses, pode ser interpretada como um limite superior para a estimativa de rentabilidade do projeto. 

## PAYBACK

O Payback nada mais é do que número de períodos necessários para que o fluxo de beneficios supere o capital investido. 

In [201]:
#Valor Presente Liquido

ativo = 140000
passivo_oneroso = 40000
patrimonio_liquido = ativo - passivo_oneroso 
ki = 0.12 #Custo emprestimo
ke = 0.15 #Custo capital proprio (custo de oportunidade ou tma)

wacc = ((passivo_oneroso/ativo)*ki)+((patrimonio_liquido/ativo)*ke)
wacc


0.1414285714285714

In [202]:
cf1 = np.array([-10000, -5000, -6000, -8000, -3000, -13000, 5000, 8000, 10000, 9000, 12000, 12500, 14000, 17000, 20000, 14000, 11000, 19000, 22000, 25000, 32000, 19000, 35000, 42000, 44000, 34000, 39000, 49000, 44000, 41000, 34000, 37500, 40000, 56000, 58000, 65000])
vpl = npf.npv(wacc, cf1)

vpl

22149.42074704751

In [203]:
#TIR - TAXA INTERNA DE RETORNO

tir = npf.irr(cf1)
tir

0.18593820085976742

In [204]:
# Valuation do negócio

lucro_operacional = 110000
despesas_capital = 50000
capital_giro = 9000

reinvestimento = (despesas_capital + capital_giro)

In [205]:
FCf = (lucro_operacional - reinvestimento)
FCf

51000

In [212]:
# Período Explícito 5 anos

vp_fcf = npf.pv(wacc, 1, pmt=-FCf, when='begin')
vp_fcf

51000.00000000002

In [213]:
# Perpetuidade

i_crescimento = 0.1

perpetuidade = (FCf*(1+i_crescimento)/(wacc-i_crescimento))
perpetuidade

1354137.931034484

In [211]:
valuation = (vp_fcf + perpetuidade)
valuation

1656097.5939935856

## L = LD + LA

onde: 
* LD =  Lucro Distribuído
* LA =  Lucro Acumulado